fix cumsum ignoring first element returned in call to promote_op

rofinn opened the pull request:
Include an option to have `promote_op` throw an error rather than return `Union{}`. We're not changing the default behaviour, but in some cases returning `Union{}` will result in less clear error messages.

Closes #53438

@LilithHafner may have comments

rofinn committed:
Added test for existing error message

rofinn committed:
Add throw_error kwarg to promote_op and update accumulate calls and tests.

rofinn commented on base/promotion.jl:
```diff
@@ -502,10 +502,12 @@ function TupleOrBottom(tt...)
 
 Guess what an appropriate container eltype would be for storing results of
 `f(::argtypes...)`. The guess is in part based on type inference, so can change any time.
+If `f(::argtypes...)` does not return (e.g., a method does not exist) then `Union{}` will be returned.
```
I could move this into the `Extended help` section if folks prefer.

LilithHafner replied:
I think this is worth mentioning in the main text, it's brief.

```suggestion
If `f(::argtypes...)` does never return (e.g., a method does not exist) then `Union{}` may be returned.
```

`promote_op` may not be clever enough to determine that f does not return. In this case it would be valid for `promote_op` to return literally any type. Consequently will => may

rofinn commented on base/accumulate.jl:
```diff
@@ -33,7 +33,7 @@ function accumulate_pairwise!(op::Op, result::AbstractVector, v::AbstractVector)
 
 function accumulate_pairwise(op, v::AbstractVector{T}) where T
-    out = similar(v, promote_op(op, T, T))
+    out = similar(v, promote_op(op, T, T; throw_error=true))
```
I'm only using `throw_error=true` in this file since that was the source of the original issue. There may be other cases where this would be useful.

rofinn commented on base/promotion.jl:
```diff
@@ -608,7 +610,13 @@ loosen the API guarantee:
 However, it is discouraged to define such unconventional API guarantees.
 """
-function promote_op(f, S::Type...)
+function promote_op(f, S::Type...; throw_error=false)
```
Default behaviour is unchanged.

rofinn commented on test/arrayops.jl:
```diff
@@ -2851,6 +2851,12 @@ end
+
+    # #53438
+    v = [(1, 2), (3, 4)]
+    @test_throws "does not return" accumulate(+, v)
```
First commit includes the original error message for this case. I can squash that if folks prefer.

jishnub commented:
The name 'throw_error` is a bit generic and doesn't indicate the purpose of the argument. I wonder if a name like 'allow_bottom` might be better?

LilithHafner commented:
Lovely! The implementation looks good. I agree that `allow_bottom` would be a better name. I'm also realizing that this encourages folks to introduce bugs like https://github.com/JuliaLang/julia/issues/53458, but I don't think that's blocking.

KristofferC commented:
For `parse` and `tryparse` there are two different functions controlling whether to throw or not. I think that makes sense here too.

rofinn commented:
Yeah, my initial thought was introducing try_promote_op, but that’s what the current behaviour does, so renaming it would be breaking? I’m open to other naming conventions. I just went with throw_error since has precedent elsewhere in the codebase.

KristofferC commented:
I'm wondering if adding a keyword argument to such a "basic" function could have some unexpected effects. This is just a worry without any supported evidence though.

vtjnash commented:
I updated the example in https://github.com/JuliaLang/julia/issues/53438 to show that this will not close that issue. I am mildly against this PR, as this is a usage bug in `accumulate` incorrectly resulting in an error here, and which seems would be even more wrong if it was using this proposed flag.

rofinn commented:
I have no strong preference here. The flag approach seemed like a relatively small change to address the `accumulate` code not handling the `Union{}` case. If we can rewrite the logic to avoid `promote_op` completely then that seems better, though more involved.

rofinn committed:
Revert "Add throw_error kwarg to promote_op and update accumulate calls and tests."

This reverts commit 06e8a349305e8814987ffcbbcb1d548bb3e26f0a.

rofinn committed:
Loosen accumulate return eltypes.

1. Moved multiple `similar` / `promote_op` calls into a single internal `_try_result` function.
2. All array calls to `_try_result` return an eltype of `promote_type(F, R)` where:
  - `F` is either `promote_op(reduce_first, ...)` or `promote_op(op, typeof(init), ...)` if an `init` kwarg is supplied
  - `R` is just `promote_op(op, eltype(array), eltype(array))`
3. Fallback to `Iterators.accumulate` if `_try_result` returns `nothing` and no `dims` kwarg is supplied.

rofinn commented:
As per @vtjnash's suggestions on issue #53438, I've reverted the `promote_op` change in favour of:

1. Consolidating the various `promote_op` calls to a single internal function in `accumulate.jl` (I have no opinion about the name)
2. Falling back to `Iterators.accumulate` if possible

NOTE: I couldn't find a way to drop the `promote_op` calls without breaking existing behaviour, so we're still using.
```
promote_type(
    isnothing(init) ? promote_op(reduce_first, typeof(op), T) : promote_op(op, typeof(init), T),
    promote_op(op, T, T)
)
```
This was also recommended in the issue thread and my understanding is that `promote_type` should fallback to a `typejoin` if no rule exists.

rofinn commented on base/accumulate.jl:
```diff
@@ -280,14 +286,25 @@ function accumulate(op, A; dims::Union{Nothing,Integer}=nothing, kw...)
+    elseif out === nothing
+        # If we can't fallback then error
+        T = eltype(A)
+        throw(ArgumentError("accumulate could not determine an appropriate container eltype for `$op($T, $T)"))
```
As far as I can tell, there isn't a way around this case?

vtjnash commented on base/accumulate.jl:
```diff
@@ -442,3 +459,22 @@ function _accumulate1!(op, B, v1, A::AbstractVector, dim::Integer)
+function _try_result(op, v::AbstractArray{T}; init=nothing) where T
+    # Try to identify the return type for the first element using `init` or `reduce_first`
+    F = isnothing(init) ? promote_op(reduce_first, typeof(op), T) : promote_op(op, typeof(init), T)
+    # Try to identify the return type fo the rest of the elements given op and eltype T
```
```suggestion
    # Try to identify the return type of the rest of the elements given op and eltype T
```

vtjnash commented on base/accumulate.jl:
```diff
@@ -442,3 +459,22 @@ function _accumulate1!(op, B, v1, A::AbstractVector, dim::Integer)
+    # Try to identify the return type for the first element using `init` or `reduce_first`
+    F = isnothing(init) ? promote_op(reduce_first, typeof(op), T) : promote_op(op, typeof(init), T)
+    # Try to identify the return type fo the rest of the elements given op and eltype T
+    R = promote_op(op, T, T)
```
```suggestion
    R = promote_op(op, F, T)
```
technically this still is wrong, since it only models the first and second arguments. The Stateful type has a similar problem in `_approx_iter_type` and you can see the approximate solution it employs there. The only place I know that handles this pattern fully complete is `@default_eltype`. I think the pattern you need here looks like this (untested):

```julia
function _accumulate_upper_bound(v, i, op, init)
    val = v[i]
    r = isnothing(init) ? reduce_first(op, val) : op(init, val)
    while inferencebarrier(nothing)
        val = v[i]
        r = op(r, val)
    end
    return r
end

function _try_result(op, v::AbstractArray{T}, init) where T
    return promote_op(_accumulate_upper_bound, typeof(v), Int, typeof(op), typeof(init))
end
```

No failure handling should be necessary (or desirable), since the only reason this returns `Union{}` is that the code will fail with the real error message before it reaches that assignment.

rofinn replied:
> No failure handling should be necessary (or desirable), since the only reason this returns `Union{}` is that the code will fail with the real error message before it reaches that assignment.

Right, thanks! Since the original issue was that `reduce_first` would work and try to assign into the array of type `Union{}` when it should have been `Tuple{Int, Int}`. If `Union{}` is returned then it should mean that all operations prior to insertion should fail.

Hmm, so the edge case your snippet is trying to address is when the return type for each call to the `op` results in us calling a different method? My understanding is that we want a "mock" function which updates a result value with the `op` over all elements of the array. We don't call it at runtime, but pass it to `promote_op` so we can ensure the widest possible type? A couple things I don't understand are:

1. What does `inferencebarrier(nothing)` do here? I thought that was used to tell the compiler to infer a value as `Any`?
2. Would this slow down inference?

In any case, yes, your example does seem to address that edge case. Here's a gist showing the difference for that edge case.

https://gist.github.com/rofinn/6aae6f559f2d9e6d1539f063ceabe2fc

vtjnash replied:
Yes, it simply tells the compile to ignore that and treat the loop condition as unknown. Alternatively, this could perhaps use eachindex there to be a bit more accurate to the iteration protocol. There shouldn't be any discernible latency cost to this, since it is just one extra method to infer behavior on

I am also also not sure that typejoin result is needed on the uniontypes of the result. It looks like can be used directly in this example.

rofinn replied:
Yeah, both work, but I was hoping to avoid overly complicated `Union` types.

rofinn replied:
Okay, two other things I changed were.

1. Switched to using `Iterators.peel` in the inference function because tests failed for arrays that didn't support linear indexing.
2. I made it a nested function to make it clear that it is only intended to be used with `promote_op`.

rofinn committed:
`_try_result` -> `similar(v, _acculuate_promote_op)`

1. `_accumulate_promote_op` just calls `promote_op` on a mock function used just for inference
2. The inference function `f` mocks applying the `op` for each element in `v` to find the widest necessary eltype
3. Updated one existing test where the return type is wider than it previously was `Vector{Union{Vector{Bool}, Vector{Int}}}`
  - This could be addressed by wrapping our result in `promote_type(uniontypes(...))`, but that will return more `Any` results.
4. Added a somewhat contrived test to demonstrate why we can't just rely in `promote_op(reduce_first, ...)` and `promote_op(op, T, T)`.
5. Also simplified some core `accumulate` logic since `_accumulate_promote_op` now handles both `init` conditions

rofinn commented on test/arrayops.jl:
```diff
@@ -2589,7 +2589,7 @@ end
     @test_throws InexactError cumsum!(similar(0x00:0xff), 0x00:0xff) # overflow
 
-    @test cumsum([[true], [true], [false]])::Vector{Vector{Int}} == [[1], [2], [2]]
+    @test cumsum([[true], [true], [false]])::Vector{Union{Vector{Bool}, Vector{Int}}} == [[1], [2], [2]]
```
This is why I was calling `promote_type(uniontypes(...))` initially.

rofinn commented on base/accumulate.jl:
```diff
@@ -442,3 +441,21 @@ function _accumulate1!(op, B, v1, A::AbstractVector, dim::Integer)
+    # NOTE: We are just passing this to promote_op for inference and should never be run.
+    function f(op, v, init)
+        val, iter = Iterators.peel(v)
+        r = isnothing(init) ? reduce_first(op, val) : op(init, val)
```
Yeah, the first option seems to give up and returns the wrong answer, but the second doesn't work because `k` needs to be defined... which is what the point of `r` is.

I really think the `promote_type(uniontypes(...)...)` has the most intuitive behaviour. It also handles the main edge case I was worried which was accidentally converting `Union{T, Missing/Nothing}` -> `Any`. Is there a particular performance issue or edge case you're concerned about?

vtjnash replied:
We don't actually run this code, so it doesn't matter if k actually gets defined

rofinn replied:
Hmm, then I'm not sure why it always return `Any`... even in simple cases like `op=+` and `v=[1, 2, 3]` 😕

rofinn replied:
Thanks, that does seem to work. So `k` does need to be defined even though the code isn't run? Is the preference for that solution simply the fact that it produces the narrowest possible answer, where `promote_union` will return `Any` when a `Union` would suffice?

vtjnash replied:
Yeah, my thought is that inference alone will generally pick a better type than `promote_union`, and that this seemed the most accurate model of what type we want it to compute

vtjnash replied:
yes, `k` needed to be defined in the right scope. I had defined it before too, but it had the wrong scope then

rofinn replied:
Sorry, I was away from my workstation for a couple weeks. The two remaining issues with the approach you suggested is that:

1. It currently doesn't address the original `Union{}` case when `op(...)` would throw an error. This happens when `k` is undefined during the `convert(promote_type(...), r)` because the `op` errors in the loop. I'm fine with adding an early exit when `k` is undefined.

2. Inference seems to give up more easily in simple cases and causes other tests to fail.
```
Expression: cumsum(Any[1, 2.3]) == [1, 3.3] == cumsum(Real[1, 2.3])::Vector{Real}
  TypeError: in typeassert, expected Vector{Real}, got a value of type Vector{Any}
```
It's less clear to me how to fix this, or whether this is an acceptable compromise.

vtjnash replied:
yes, there could be an `isdefined` check there, but you have to be careful of how to write that because when written statically that check would always be true and result in skipping computing the desired promotion.

My other thought was that it might be possible to split `f` into 2 parts (before and after computing `k`) with the first one computing `typeof(k)`, then calling this `promote_type` directly (instead of inferring it)
```
        # next try to change `r` to that type
        r = convert(promote_type(typeof(r), typeof(k)), r)
```

My main concern with using `Base.promote_union` is that it can discard a lot of information and be quite terrible at computing the output if you have any unions in the input array. Whereas `promote_op` originally should be able to give a precise answer in those cases, which `promote_union` then discards.

rofinn replied:
Hmm, it's not pretty, but it looks like breaking up `f` works.

1. Avoids the `isdefined` check
2. Fixes the test where inference was giving up and returning `Any`.

I'll push a commit, so you can take a look. Maybe there's a cleaner way to right it...?

rofinn committed:
Use promote_union to fix previous broken test

1. All previous tests pass now
2. Union{Int, Float64, ComplexF64, String} -> Any
3. Specific Union types like Union{T, Missing/Nothing} work as expected

NOTE: promote_type(uniontypes(T)...) did hit some `@inferred` issues though.

rofinn committed:
Replace promote_union with a dual-pass promote_op solution.

It isn't pretty, but all existing tests still pass now and it is more specific in one test case.
https://github.com/JuliaLang/julia/pull/53461#discussion_r1508233061

vtjnash reviewed (approved):
I think this looks great
